libsql-server/src/namespace/configurator/primary.rs

```diff
@@ -4,125 +4,192 @@ use std::{path::Path, pin::Pin, sync::Arc};
 use crate::connection::MakeConnection;
 use crate::database::{Database, PrimaryDatabase};
-use crate::namespace::{Namespace, NamespaceConfig, NamespaceName, NamespaceStore, ResetCb, ResolveNamespacePathFn, RestoreOption};
-use crate::namespace::meta_store::MetaStoreHandle;
 use crate::namespace::broadcasters::BroadcasterHandle;
+use crate::namespace::configurator::helpers::make_primary_connection_maker;
+use crate::namespace::meta_store::MetaStoreHandle;
+use crate::namespace::{
+    Namespace, NamespaceBottomlessDbIdInit, NamespaceName, NamespaceStore,
+    ResetCb, ResolveNamespacePathFn, RestoreOption,
+};
 use crate::run_periodic_checkpoint;
 use crate::schema::{has_pending_migration_task, setup_migration_table};
 
-use super::ConfigureNamespace;
+use super::helpers::cleanup_primary;
+use super::{BaseNamespaceConfig, ConfigureNamespace, PrimaryExtraConfig};
+
+pub struct PrimaryConfigurator {
+    base: BaseNamespaceConfig,
+    primary_config: PrimaryExtraConfig,
+    make_wal_manager: Arc<dyn Fn() -> InnerWalManager + Sync + Send + 'static>,
+}
+
+impl PrimaryConfigurator {
+    pub fn new(
+        base: BaseNamespaceConfig,
+        primary_config: PrimaryExtraConfig,
+        make_wal_manager: Arc<dyn Fn() -> InnerWalManager + Sync + Send + 'static>,
+    ) -> Self {
+        Self {
+            base,
+            primary_config,
+            make_wal_manager,
+        }
+    }
+
+    #[tracing::instrument(skip_all, fields(namespace))]
+    async fn try_new_primary(
+        &self,
+        namespace: NamespaceName,
+        meta_store_handle: MetaStoreHandle,
+        restore_option: RestoreOption,
+        resolve_attach_path: ResolveNamespacePathFn,
+        db_path: Arc<Path>,
+        broadcaster: BroadcasterHandle,
+    ) -> crate::Result<Namespace> {
+        let mut join_set = JoinSet::new();
+
+        tokio::fs::create_dir_all(&db_path).await?;
+
+        let block_writes = Arc::new(AtomicBool::new(false));
+        let (connection_maker, wal_wrapper, stats) = make_primary_connection_maker(
+            &self.primary_config,
+            &self.base,
+            &meta_store_handle,
+            &db_path,
+            &namespace,
+            restore_option,
+            block_writes.clone(),
+            &mut join_set,
+            resolve_attach_path,
+            broadcaster,
+            self.make_wal_manager.clone(),
+        )
+        .await?;
+        let connection_maker = Arc::new(connection_maker);
+
+        if meta_store_handle.get().shared_schema_name.is_some() {
+            let block_writes = block_writes.clone();
+            let conn = connection_maker.create().await?;
+            tokio::task::spawn_blocking(move || {
+                conn.with_raw(|conn| -> crate::Result<()> {
+                    setup_migration_table(conn)?;
+                    if has_pending_migration_task(conn)? {
+                        block_writes.store(true, Ordering::SeqCst);
```

Comment:
if this fails do we just keep it blocked forever? Is that going to be a problem?

Reply:
this is not new, just moved over here. if that fails, the namespace won't be loaded, so we assume that there will be a retry. I that keeps failing forever then yes, but that's another problem (like out of disk or some stuff like that)